### ISSUE TITLE
Feat: Provide mechanism to register onCancel and onClose handlers

### ISSUE THREAD

**Issue by mattp-swirldslabs** [hashgraph/pbj#303]
### Problem

As an application developer using the helidon plugin framework
I want a mechanism to register Runnable callbacks for onCancel and onClose events
So that I can deallocate resources when a client closes or cancels a connection

### Solution

Provide a registration mechanism like `grpc.io`'s `setOnCancelHandler()` and `setOnCloseHandler()` so the application can register Runnable callbacks.

**Comment by mattp-swirldslabs** [comment]
A candidate for where cancel events could invoke a registered Runnable: https://github.com/hashgraph/pbj/blob/845c93d74c280401fb4aa4f3253cd04b03d519f3/pbj-core/pbj-runtime/src/main/java/com/hedera/pbj/runtime/grpc/Pipelines.java#L790-L792

**Comment by mattp-swirldslabs** [comment]
Example of how the registered callback is invoked in ServerCalls in `grpc.io`:
<img width="1433" alt="image" src="https://github.com/user-attachments/assets/6552b7e4-e793-4e19-8e4d-6b48b8f48f8d">

**Comment by litt3** [comment]
@mattp-swirldslabs A couple clarifications:

- Is there a specific need to have both `onCancel` *and* `onClose` callbacks?
- When you  refer to a client canceling, do you mean sending an `RST_STREAM` frame, or are there other ways to "cancel"?
- Looking at the case of a client canceling: are you imagining both the `onCancel` *and* the `onClose` callbacks running? Or would it be either-or?

**Comment by rbair23** [comment]
I don't understand the ask. Since at this time PBJ is not generating the grpc service endpoints, but these are created by the application, you have access to cancel and close using the Java Flow APIs? Why do we need to add something, and where would it be added?

**Comment by rbair23** [comment]
Is it on the UnaryBuilder and other builders that you want these methods added?

**Comment by rbair23** [comment]
It should be that PBJ auto-generates service interfaces, such as [in this example (GreeterService.java)](https://github.com/hashgraph/pbj/blob/845c93d74c280401fb4aa4f3253cd04b03d519f3/pbj-core/pbj-grpc-helidon/src/test/java/com/hedera/pbj/grpc/helidon/GreeterService.java) And then the application would implement this interface, such as [in this test](https://github.com/hashgraph/pbj/blob/845c93d74c280401fb4aa4f3253cd04b03d519f3/pbj-core/pbj-grpc-helidon/src/test/java/com/hedera/pbj/grpc/helidon/PbjTest.java#L793).

There are 4 types of methods:
- Unary
- Client-Streaming
- Server-Streaming
- Bidirectional-Streaming

Unary requests have no streams, and therefore no Flow. They just take the protobuf request object as a Java record, and return a protobuf response object as a Java record.

Client-Streaming is where the client is going to send a whole stream of objects, and in this case, the server needs to know when the connection is terminated so it can clean up resources.

Server-Streaming is where the server is streaming results to the client. In this case, the server could benefit from some callback _or exception_ that indicates the client has closed the connection. An exception in this case might be more useful.

Bidirection-Streaming has to worry about both cases -- the client closing the stream being sent to the server, and the client closing the connection being sent _from_ the server.

Suppose you are implementing `GreeterService.java` and your method signature is:
```
public Flow.Subscriber<? super HelloRequest> sayHelloStreamRequest(
                Flow.Subscriber<? super HelloReply> replies) {
            // ...
        }
```

The `replies` object is the one you will push replies to, and the `Flow.Subscriber` that you return is the one the client will send requests to you on. This is an example of the "bidirectional-streaming" case. But actually the same method signature is used for the server-streaming case. So I would say, this is oddity #1: the server-streaming case should maybe be given a Consumer style lambda instead of a Flow, since it is only expected to send a single response.

In either client-streaming or bidirection-streaming, the `Flow.Subscriber` returned from this method is one that will be given requests from the client over and over. This flow object has the following methods:

```
            return new Flow.Subscriber<>() {
                @Override
                public void onSubscribe(Flow.Subscription subscription) {
                    subscription.request(Long.MAX_VALUE); // turn off flow control
                }

                @Override
                public void onNext(HelloRequest item) {
                    // ...
                }

                @Override
                public void onError(Throwable throwable) {
                    // ...
                }

                @Override
                public void onComplete() {
                    // ...
                }
            };
```

If the client cancels the request, or completes it, the appropriate override method is called. So we already have a mechanism by which the server handler knows when the client has been closed.

For the responses, I think there is a gap, where during the response unless some exception is thrown, a server could send a stream of responses to a broken client connection. And I don't remember any method by which the server can be told "hey, the client is gone, you should stop sending". Maybe an exception is the most reasonable, but the Flow API doesn't provide for it. We could subclass the Flow.Subscriber API to add a "throws" clause, or documentation, that says onNext might through some exception if the connection is closed, allowing the server code to use a try/catch block and clean up resources.

So in summary, I think we already have the mechanism asked for when it comes to client connections closing _on the request side_, but maybe not when the connection is closed _on the response side_, which impacts server-streaming the most (bidirectional-streaming should get a notice when the client connection closes).

**Comment by mattp-swirldslabs** [comment]
Thanks for the comments. I've been trying different scenarios with `bidi` and `server streaming`. @rbair23 as you noted with this comment:

> For the responses, I think there is a gap, where during the response unless some exception is thrown, a server could send a stream of responses to a broken client connection. And I don't remember any method by which the server can be told "hey, the client is gone, you should stop sending"

If a consumer is streaming response with a `server streaming` connection and they drop the connection, a SocketException is thrown in `PbjProtocolHandler.onNext()`. Right now this just fills the logs with error messages. Calling something like, `pipeline.error()` does not work because it invokes the `SendToClientSubscriber.error()`. Instead, it would be helpful if I can register a Runnable on the `Flow.Subscriber<? super SubscribeStreamResponse>` passed when the application code is first invoked via `Pipelines.serverStreaming().method()`.

I think there's a similar issue in `bidi` connections if the socket times out because the producer is not sending any data. A SocketException is thrown and we need to set a Runnable to be invoked when the exception is caught.

**Comment by mattp-swirldslabs** [comment]
<img width="917" alt="image" src="https://github.com/user-attachments/assets/666434b3-673b-43f9-9944-d86fd7570a41">

**Comment by mattp-swirldslabs** [comment]
<img width="891" alt="image" src="https://github.com/user-attachments/assets/7d91b5ab-2bba-4617-9e53-1fde15f9862d">

**Comment by anthony-swirldslabs** [comment]
If I understand the problem statement correctly, it concerns the case when the server is streaming responses to the client, whether in the `Server-Streaming` or `Bidirectional-Streaming` mode. And the ask is to enable the server to detect the situation when the client is unable to continue to receive any more responses from the server (e.g. due to the connection being closed for whatever reason), so that the server could stop sending the responses and clean up resources.

I see that @rbair23 suggests that an exception thrown from the `replies.onNext()` can be used for this naturally. I tend to agree with this suggestion as this solution doesn't add any additional complexity to the implementation in PBJ. Further, it seems to simplify the custom server implementation as well because the server application wouldn't need to implement any additional communication between an "onCancel"/"onClose" `Runnable` callback and the actual method that is sending the responses - the exception would instead pop up directly in this very method, at the very time when this situation occurs, thus allowing the server to react immediately as appropriate.

Finally, I see that the current implementation in PBJ already throws an exception when a network error (or serialization, or any other failure) occurs:

https://github.com/hashgraph/pbj/blob/5865614c9d0d2ff0973b6ba45af126a78fd8e8ff/pbj-core/pbj-grpc-helidon/src/main/java/com/hedera/pbj/grpc/helidon/PbjProtocolHandler.java#L667-L671

So I believe that there's not a lot of scope for any additional work that PBJ needs to do to support the feature request as it seems to be fully supported already.

I don't think that adding a `throws` clause with a checked exception to the `onNext()` method would be a good idea because it would require one to handle the exception everywhere, which may not be super necessary in certain cases (e.g. when there's not any resources to clean up.) However, we do have a javadoc `@throws` note at:

https://github.com/hashgraph/pbj/blob/5865614c9d0d2ff0973b6ba45af126a78fd8e8ff/pbj-core/pbj-runtime/src/main/java/com/hedera/pbj/runtime/grpc/Pipeline.java#L19-L19

which is sufficient for the purpose of documenting the behavior. The author of a server implementation can decide whether to handle the exception or let it propagate up the stack.

So I think we can close the issue. @rbair23 , @jsync-swirlds any comments?

**Comment by jsync-swirlds** [comment]
Personally, I would prefer to see a checked exception, or at minimum a subclass of `RuntimeException` rather than a raw exception.

I may be missing something, but in the `server-streaming` case (which is actually the harder of the two), the code seems to generate a no-op for client-end-streaming, which I would think is related to a client ending the stream...

I wasn't part of the original testing (and the code that _was_ using this plugin is not sticking around), so we might have to reset and retry some things after the current "whack it with a hammer and rebuild" work on the block node is complete to see if this works as described or if there is something still needed.

**Comment by anthony-swirldslabs** [comment]
Thanks for the feedback. I don't mind introducing a custom exception inheriting from the `RuntimeException`. I'm a bit reluctant to checked exceptions as mentioned above, but if more people strongly prefer them, we can consider that as well.

Also, as described above, if I read the code correctly, I believe that the moment the client closes the connection to which the server is streaming responses, the exception mentioned above will be thrown which the server code should be able to detect. I'm unsure I understand the `client-end-streaming` in the context of the `server-streaming` case though - there's not any streams from the client in this case. The client shoots a single request at the server and then it listens to the stream of responses from the server. The only connection that matters for the server to continue streaming is the latter - as long as it remains open, it indicates that the client is still listening to the stream and consuming all the data, so the server should keep streaming until the connection is closed/terminated (or the request has finished processing.)

However, since all the code that uses this API is currently being rewritten, let's indeed wait and see if this issue is still relevant and what changes we actually need to make. For now, I'll keep this ticket in the `Blocked` state until/unless we hear in the comments.

**Comment by Nana-EC** [comment]
@jsync-swirlds can you take a look here.
Is this still applicable and if so is it Preview BN/MVP applicable?
